### PR TITLE
Added AddPerformanceMaxProductListingGroupTree example

### PR DESCRIPTION
**Pull request by fiboknacky**
This is based on https://github.com/googleads/google-ads-dotnet/blob/main/examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.cs.

But I don't use the factory pattern like in the .NET library, as it would clutter the implementation.
The current version should be close to the AddShoppingProductListingGroupTree example, so the users should be familiar with this implementation.

### CODE REVIEW

**Comment by Kenimartonia** [comment]
Thanks and best regards

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+        foreach ($response->iterateAllElements() as $googleAdsRow) {
+            /** @var GoogleAdsRow $googleAdsRow */
+            $assetGroupListingGroupFilters[] = $googleAdsRow->getAssetGroupListingGroupFilter();
+        }
+        return $assetGroupListingGroupFilters;
+    }
+    // [END add_performance_max_product_listing_group_tree_7]
+
+    /**
+     * Removes an existing tree of asset group listing group filters.
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:312]: It does not remove them, it only creates the mutate operations to do so.

Please replace with `Creates mutate operations for removing ...`

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+     *
+     * Asset group listing group filters must be removed in a specific order: all of the children
+     * of a filter must be removed before the filter itself, otherwise the API will return an
+     * error.
+     *
+     * @param GoogleAdsClient $googleAdsClient the Google Ads API client
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:318]: Both the parameters `$googleAdsClient` and `$customerId` do not exist.

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+     * @param int $customerId the customer ID
+     * @param AssetGroupListingGroupFilter[] $assetGroupListingGroupFilters the existing asset
+     *     group listing group filters
+     * @return MutateOperation[] the list of MutateOperations to remove all listing groups
+     */
+    private static function createMutateOperationsForRemovingListingGroupFiltersTree(
+        array $assetGroupListingGroupFilters
+    ): array {
+        if (empty($assetGroupListingGroupFilters)) {
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:327]: This seems unnecessary (we only call this method when there are in the main method).

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+            throw new \UnexpectedValueException('No listing group filters to remove');
+        }
+
+        $resourcesToListingGroupFilters = [];
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:331]: `resourceNamesToListingGroupFilters`

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+            $parentResourceName = $assetGroupListingGroupFilter->getParentListingGroupFilter();
+            $siblings = [];
+
+            // Checks to see if we've already visited a sibling in this group, and fetches or
+            // creates a new set as required.
+            if (array_key_exists($parentResourceName, $parentsToChildren)) {
+                $siblings = $parentsToChildren[$parentResourceName];
+            }
+            $siblings[] = $assetGroupListingGroupFilter->getResourceName();
+            $parentsToChildren[$parentResourceName] = $siblings;
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:356]: This is only necessary when executed in an `else` clause of the condition above

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+     * factory's asset group.
+     *
+     * The root node or partition is the default, which is displayed as "All Products".
+     *
+     * @param int $customerId the customer ID
+     * @param int $assetGroupId the asset group ID
+     * @param int $rootListingGroupId the root listing group ID
+     * @return MutateOperation the mutate operation for creating the root
+     */
+    // [START add_performance_max_product_listing_group_tree_4]
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:415]: The golden includes the function comment (not sure if intended but let's keep it consistent)

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+    private static function createMutateOperationForRoot(
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:416]: This function seems to be unnecessary, should we refactor to remove it altogether?

If we want to keep it, I suggest duplicating the content of `createMutateOperationForSubdivision` and removing the support of the special case for root in `createMutateOperationForSubdivision` for easier understanding.

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+     * @param ListingGroupFilterDimension|null $listingGroupFilterDimension the listing group
+     *     filter dimension to associate with the asset group listing group filter
+     * @return MutateOperation the mutate operation for creating a subdivision
+     */
+    // [START add_performance_max_product_listing_group_tree_5]
+    private static function createMutateOperationForSubdivision(
+        int $customerId,
+        int $assetGroupId,
+        int $assetGroupListingGroupFilterId,
+        ?int $parentId = null,
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:447]: `= null` can be removed (`?` is enough). Same thing for the next parameter in the list.

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+     * @param int $assetGroupId the asset group ID
+     * @param int $assetGroupListingGroupFilterId the ID of the asset group listing group filter to
+     *     be created
+     * @param int|null $parentId the ID of the parent of asset group listing group filter to be
+     *      created
+     * @param ListingGroupFilterDimension|null $listingGroupFilterDimension the listing group
+     *     filter dimension to associate with the asset group listing group filter
+     * @return MutateOperation the mutate operation for creating a subdivision
+     */
+    // [START add_performance_max_product_listing_group_tree_5]
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:442]: ditto: The golden includes the function comment (not sure if intended but let's keep it consistent)

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 10]
```diff
@@ -0,0 +1,602 @@
...
+     * @param int $assetGroupId the asset group ID
+     * @param int $assetGroupListingGroupFilterId the ID of the asset group listing group filter to
+     *     be created
+     * @param int $parentId the ID of the parent of asset group listing group filter to be
+     *      created
+     * @param ListingGroupFilterDimension $listingGroupFilterDimension the listing group
+     *     filter dimension to associate with the asset group listing group filter
+     * @return MutateOperation the mutate operation for creating a unit
+     */
+    // [START add_performance_max_product_listing_group_tree_6]
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:504]: ditto: The golden includes the function comment (not sure if intended but let's keep it consistent)

**Review by PierrickVoulet — CHANGES_REQUESTED** [review]

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 11]
```diff
@@ -0,0 +1,602 @@
...
+            'SELECT asset_group_listing_group_filter.resource_name, '
+            . 'asset_group_listing_group_filter.parent_listing_group_filter '
+            . 'FROM asset_group_listing_group_filter '
+            . 'WHERE asset_group_listing_group_filter.asset_group = "%s"',
+            $assetGroupResourceName
+        );
+
+        // Issues a search request by specifying page size.
+        $response =
+            $googleAdsServiceClient->search($customerId, $query, ['pageSize' => self::PAGE_SIZE]);
```
[the remark this answers is not in the document]
Comment (fiboknacky) [reply examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:296]: I think it's fine for both ways. The golden uses paging so I follow it for consistency.

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 12]
```diff
@@ -0,0 +1,602 @@
...
+     * and uses it to extract the new entity's name and resource name.
+     *
+     * @param MutateOperation[] $mutateOperations the submitted mutate operations
+     * @param MutateGoogleAdsResponse $mutateGoogleAdsResponse the mutate Google Ads response
+     */
+    private static function printResponseDetails(
+        array $mutateOperations,
+        MutateGoogleAdsResponse $mutateGoogleAdsResponse
+    ): void {
+        $i = 0;
```
[the remark this answers is not in the document]
Comment (fiboknacky) [reply examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:553]: Done. I need to change the variable name a bit, or I cannot make it pass PSR-12 without having line length over 100 chars.

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 13]
```diff
@@ -0,0 +1,602 @@
...
+    // [END add_performance_max_product_listing_group_tree_6]
+
+    /**
+     * Prints the details of a mutate google ads response. Parses the "response" oneof field name
+     * and uses it to extract the new entity's name and resource name.
+     *
+     * @param MutateOperation[] $mutateOperations the submitted mutate operations
+     * @param MutateGoogleAdsResponse $mutateGoogleAdsResponse the mutate Google Ads response
+     */
+    private static function printResponseDetails(
```
[the remark this answers is not in the document]
Comment (fiboknacky) [reply examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:549]: I think you need to weigh that with consistency. The golden CL does this.
[AddPerformanceMaxCampaign](https://github.com/googleads/google-ads-php/blob/main/examples/AdvancedOperations/AddPerformanceMaxCampaign.php#L648) also does something similar to this file, though it's a bit simpler.

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php (hunk 5 above)
Comment (fiboknacky) [reply examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:356]: I don't think so. Can you tell why you think like that?

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php [hunk 14]
```diff
@@ -412,40 +405,54 @@ private static function createMutateOperationsForRemovingDescendents(
...
+            // another sub-tree.
+            'type' => ListingGroupFilterType::SUBDIVISION,
+            // Because this is a Performance Max campaign for retail, we need to specify
+            // that this is in the shopping vertical.
+            'vertical' => ListingGroupFilterVertical::SHOPPING
+        ]);
+
+        return new MutateOperation([
+            'asset_group_listing_group_filter_operation'
+            => new AssetGroupListingGroupFilterOperation([
```
Comment (PierrickVoulet) [inline examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:430]: Indentation is not perfect.

File: examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php (hunk 5 above)
Comment (PierrickVoulet) [reply examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php:348]: It works but I find the logic a bit confusing.

I would expect that we only have to execute `$parentsToChildren[$parentResourceName] = $siblings` with `$siblings` a new array when `$parentsToChildren[$parentResourceName]` is unset.

I guess is it nit-picking so we can keep it as-is.

**Review by PierrickVoulet — CHANGES_REQUESTED** [review]

**Review by PierrickVoulet — APPROVED** [review]

**Review by wfansh — APPROVED** [review]

**Pull request merged**